rpc/internal/collector.service.go

```diff
@@ -190,54 +205,107 @@ func (sc *streamsCoordinator) getStream(name string) (string, *streamGroup, erro
 	return id, group, nil
 }
 
-// addEvent writes the given event to the collector from the given stream. If
-// the stream does not exist an error is returned. Note that this function
-// blocks until all streams in the group have an entry in the heap, at which
-// point the timestamp can be guaranteed.
+// addEvent writes the given event to the stream's buffer. If the time since
+// the last flush has surpassed the flush interval, the flush method is called.
+// If the stream does not exist or is already closed an error is returned.
 func (sg *streamGroup) addEvent(ctx context.Context, id string, event *events.Performance) error {
 	sg.mu.Lock()
+	defer sg.mu.Unlock()
+
+	if sg.catcher.HasErrors() {
+		return sg.catcher.Resolve()
+	}
 
-	errChan, ok := sg.streams[id]
+	stream, ok := sg.streams[id]
 	if !ok {
-		sg.mu.Unlock()
-		return errors.Errorf("stream %s does not exist in this stream group", id)
+		return errors.Errorf("stream '%s' does not exist in this stream group", id)
+	}
+	if stream.closed {
+		return errors.Errorf("stream '%s' already closed", id)
 	}
 
-	sg.eventHeap.SafePush(&performanceHeapItem{errChan: errChan, event: event})
-	if sg.eventHeap.Len() >= len(sg.streams) {
-		item := sg.eventHeap.SafePop()
-		go sg.sendError(item)
+	if stream.inHeap {
+		stream.buffer <- event
+		return nil
 	}
 
-	sg.mu.Unlock()
+	var e *events.Performance
 	select {
-	case <-ctx.Done():
-		return errors.Wrap(ctx.Err(), "context canceled while adding event")
-	case err := <-errChan:
-		return err
+	case e = <-stream.buffer:
+		stream.buffer <- e
+	default:
+		e = event
 	}
-}
+	sg.eventHeap.SafePush(&performanceHeapItem{id: id, event: event})
+	stream.inHeap = true
 
-// sendError writes the next item from the heap to the collector and sends the
-// error, if any, to the corresponding error channel.
-func (sg *streamGroup) sendError(item *performanceHeapItem) {
-	if item != nil {
-		item.errChan <- sg.collector.AddEvent(item.event)
+	if time.Since(sg.lastFlush) > 5*time.Second { // TODO: figure out a good flush interval
```

Comment:
Is flushing on an interval necessary for performance reasons? Or is there any reason we can't simply flush as soon as we know the event heap has enough items in it?

Reply:
Performance reasons, if we flush every time this becomes blocking because it can potentially hold the lock for a while depending on the size of all the stream buffers.